src/werkzeug/datastructures/structures.py

```diff
@@ -88,6 +88,54 @@ def get(self, key, default=None, type=None):
                 rv = default
         return rv
 
+    def pop(self, key, default=_missing, type=None):
+        """Like :meth:`get` but removes the key/value pair.
+
+        >>> d = TypeConversionDict(foo='42', bar='blub')
+        >>> d.pop('foo', type=int)
+        42
+        >>> 'foo' in d
+        False
+        >>> d.pop('bar', -1, type=int)
+        -1
+        >>> 'bar' in d
+        False
+
+        :param key: The key to be looked up.
+        :param default: The default value to be returned if the key is not
+                        in the dictionary. If not further specified it's
+                        an :exc:`KeyError`.
+        :param type: A callable that is used to cast the value in the dict.
+                        If a :exc:`ValueError` or a :exc:`TypeError` is raised
+                        by this callable the default value is returned.
+
+        .. admonition:: note
+
+           If the type conversion fails, the key is **not** removed from the
+           dictionary.
+
+        """
+        try:
+            rv = self[key]
+        except KeyError:
+            if default is _missing:
+                raise
+            return default
+        if type is not None:
+            try:
+                rv = type(rv)
+            except (ValueError, TypeError):
+                if default is _missing:
+                    return None
```

Comment:
Above, we raise the error if the default is missing. Seems like it should be raised here too, not silently succeed.

Reply:
It also doesn't match the docstring which says "if type conversion fails the item is not removed"

Reply:
I pushed an updated version with this change. Leaving this open for discussion.

Reply:
After thinking a bit more, I still think raising the error makes sense. `dict.pop(key)` will raise an error if the key doesn't exist, `dict.pop(key, None)` can be used to silence that error. Extending that to "key exists but conversion fails" makes sense.